### DISCUSSION TITLE
Named field access in custom element hooks

### DISCUSSION THREAD

**Discussion by okomestudio** [minad/tempel#152] [Ideas]
I propose that the function hooks for custom elements accept an extra argument `ST`, so that it has access to named field values for the current template.

The motivation for the change follows.

Currently, the function hook has the signature

```elisp
(defun my-elt (elt) ...)
```

for which the proposed change results in

```elisp
(defun my-elt (elt st) ...)
```

so that named fields will be available as an alist via `(cdr st)`. This change appears necessary to access named fields, as writing a template like

```elisp
(my-template (p "my-field: " my-field t) (my-elt my-field))
```

does not pass the value of `my-field` as the `elt` argument to `my-elt`.

On the other hand, if there is a better way to achieve named field access, the proposed change will not be necessary. In that case, I’d like to know what the available method is.

This may be classified as a breaking change, and if accepted, the README section on the topic will need an update. A proof-of-concept PR will follow.

**Comment by minad** [comment]
`tempel-user-elements` is modeled after `tempo-user-element-functions` so the API is kind of fixed. Also I don't want to expose the internal state, since the structure may change. Would it suffice to only pass `(cdr st)`? This would be a little more robust I think.

Ideally the change would be backward compatible such that user elements with and without argument would work. This can be achieved via `run-hook-wrapped` and `condition-case` catching call arity errors.

**Reply by okomestudio** [reply]
I understand the need for API alignment. So, what you mean is along the line of this?

```elisp
(defun tempel--user-elements-wrapper (hook elt fields)
  "Wrap HOOK with correct number of arguments, ELT and FIELDS."
  (condition-case e
      (funcall hook elt)
    (wrong-number-of-arguments
     (funcall hook elt fields))))

(defun tempel--element (st region elt)
  ...
    (_ (if-let ((ret (run-hook-wrapped 'tempel-user-elements
                                       #'tempel--user-elements-wrapper
                                       elt (cdr st))))
         ...)
```

This works for me. And it does make sense to maintain backward compatibility.

**Reply by minad** [reply]
Yes, I am happy to accept a PR which adds this. But please use a lambda instead of tempel--user-elements-wrapper and adapt the documentation accordingly.

**Reply by okomestudio** [reply]
Will submit a PR tomorrow with the suggested change. Thank you for your quick feedback.

**Reply by okomestudio** [reply]
PR has been created: https://github.com/minad/tempel/pull/153
